pkgs/by-name/ev/evaluated/evaluator.nix

```diff
@@ -0,0 +1,40 @@
+let
+
+  pkgs = import <nixpkgs> { };
+  inherit (pkgs.lib)
+    concatStringsSep
+    isDerivation
+    isFunction
+    isAttrs
+    mapAttrs
+    ;
+  /**
+    Helper for package tree introspection.
+
+    It emits the position of all functions
+    and derivations using `builtins.trace`
+    so that another program can quickly calculate
+    the difference of packages between two commits.
+  */
+  traceTree =
+    let
+      traceTree' =
+        tree: path:
+        let
+          pathStr = concatStringsSep "." path;
+          evaluated = builtins.tryEval tree;
+        in
+        if isFunction tree then
+          builtins.trace ":tree: function ${pathStr} :tree:" null
+        else if !evaluated.success then
+          builtins.trace ":tree: exception ${pathStr} :tree:" null
+        else if isDerivation tree then
+          builtins.trace (builtins.unsafeDiscardStringContext ":tree: derivation ${pathStr} ${tree.drvPath} :tree:") null
```

Comment:
How does it compare to the existing approach using nix-env. That one seems to be faster in our current CI than the 10 minutes that you are describing?

Reply:
This is what we also do in nixpkgs-review.

Reply:
In this case we can also diff added functions, not only derivations

IDK exactly how nix-env traverses packages so I don't know the differences in approach

Reply:
What would we do with the function diff? We cannot test those directly, can we? Should be added function not also obvious from reading the diff?

Reply:
Here is what is currently done: https://github.com/NixOS/nixpkgs/blob/9bd1d5605fe165b2b247f0d2113fde590f393106/ci/eval/default.nix#L86

Reply:
> What would we do with the function diff? We cannot test those directly, can we? Should be added function not also obvious from reading the diff?

Can't test, yet, but it shows in the diff when another function is introduced. Maybe when we introduce some kind of `passthru.tests` for functions we could automate testing only the functions that changed and it's dependents.

BTW my intention here is to experiment with a different approach, like how faster or slower can we go if we builtins.trace the stuff instead of returning a data structure using nix-env, and running it on a derivation so it could be offloaded to a remote builder.

Maybe the implementation could be faster but I don't know how to optimize further.

Reply:
Sure. I just wanted of how things work today. I also did some experimenting in the past with walking the tree with nix, but wasn't able to beat the C++ function that nix-env uses internally yet. Also this new github eval goes in a similar direction.